core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala

```diff
@@ -0,0 +1,278 @@
+
+/**
+ * Physical plan node for scanning data from a SnappyData column table RDD.
+ * If user knows that the data is partitioned across all nodes, this SparkPlan
+ * can be used to avoid expensive shuffle and Broadcast joins.
+ * This plan overrides outputPartitioning and makes it inline with the
+ * partitioning of the underlying DataSource.
+ */
+private[sql] final case class ColumnTableScan(
+    output: Seq[Attribute],
+    rdd: RDD[Any],
+    otherRDDs: Seq[RDD[InternalRow]],
+    numPartitions: Int,
+    numBuckets: Int,
+    partitionColumns: Seq[Expression],
+    scanUnsafeRows: Boolean,
+    @transient baseRelation: PartitionedDataSourceScan)
+    extends PartitionedPhysicalRDD(output, rdd, numPartitions, numBuckets,
+      partitionColumns, baseRelation) {
+
+  private[sql] override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext,
+      "number of output rows"),
+    "numRowBufferOutput" -> SQLMetrics.createMetric(sparkContext,
+      "number of output rows from the row buffer")) ++ (
+      if (otherRDDs.isEmpty) Map.empty
+      else Map("numOtherRows" -> SQLMetrics.createMetric(sparkContext,
+        "number of output rows from other RDDs")))
+
+  private val allRDDs = if (otherRDDs.isEmpty) rdd
+  else rdd.sparkContext.union((Seq(rdd) ++ otherRDDs).asInstanceOf[Seq[RDD[Any]]])
+
+  private val internalRowPartitionIndex = rdd.getNumPartitions
+
+  override def inputRDDs(): Seq[RDD[InternalRow]] = {
+    allRDDs.asInstanceOf[RDD[InternalRow]] :: Nil
+  }
+
+  override def doProduce(ctx: CodegenContext): String = {
+    val numOutputRows = metricTerm(ctx, "numOutputRows")
+    val numRowBufferOutput = metricTerm(ctx, "numRowBufferOutput")
+    val numOtherRows =
+      if (otherRDDs.isEmpty) null else metricTerm(ctx, "numOtherRows")
+    val numRows = ctx.freshName("numRows")
+    // PartitionedPhysicalRDD always has one input.
+    // It returns an iterator of iterators (row + column)
+    // except when doing union with multiple RDDs where other
+    // RDDs return iterator of UnsafeRows.
+    val rowInput = ctx.freshName("rowInput")
+    val colInput = ctx.freshName("colInput")
+    if (otherRDDs.isEmpty) {
+      ctx.addMutableState("scala.collection.Iterator",
+        rowInput, s"$rowInput = (scala.collection.Iterator)inputs[0].next();")
+      ctx.addMutableState("scala.collection.Iterator",
+        colInput, s"$colInput = (scala.collection.Iterator)inputs[0].next();")
+    } else {
+      ctx.addMutableState("scala.collection.Iterator",
+        rowInput,
+        s"""
+            $rowInput = (partitionIndex < $internalRowPartitionIndex)
+                ? (scala.collection.Iterator)inputs[0].next() : inputs[0];
+        """)
+      ctx.addMutableState("scala.collection.Iterator",
+        colInput,
+        s"""
+            $colInput = (partitionIndex < $internalRowPartitionIndex)
+                ? (scala.collection.Iterator)inputs[0].next() : null;
+        """)
```

Comment:
As I understand, partitionIndex for the otherRDDs should be beyond the total number of rdd partitions of column table. Its quite novel (and may be performant) to bring the iteration of three rdds in a single function. However, can this dependency of having otherRDDs partitionindex greater than the column table be avoided. It will be difficult to understand and maintain this later.